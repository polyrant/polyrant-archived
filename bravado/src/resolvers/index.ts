export { PostResolver } from './post';
export { PingResolver } from './ping';
