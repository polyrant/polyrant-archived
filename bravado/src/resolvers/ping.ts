export default {
  Query: {
    ping: async (): Promise<string> => {
      return 'Pong!';
    },
  },
};
