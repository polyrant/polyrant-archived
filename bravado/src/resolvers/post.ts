import { Post } from '../models';
import { Args } from '../types';

import { IResolvers } from 'apollo-server-express';

const resolvers: IResolvers = {
  /************************* Queries *************************/

  Query: {
    // Get all posts
    posts: () => {
      return Post.find();
    },

    // Get a post
    post: (_: void, { id }: Args) => {
      return Post.findOne({ id });
    },
  },

  /************************* Mutations *************************/

  Mutation: {
    // Create a post
    createPost: async (_: void, { content }: Args) => {
      const post = Post.create({ content });
      return post;
    },

    // Delete a post
    deletePost: async (_: void, { id }: Args): Promise<boolean> => {
      try {
        Post.findOneAndDelete({ id });
        return true;
      } catch (err) {
        return false;
      }
    },
  },

  /*************************  *************************/
};

export default resolvers;
