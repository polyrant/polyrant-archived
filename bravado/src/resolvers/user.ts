import { IResolvers } from 'apollo-server-express';
import argon2 from 'argon2';

import { User } from '../entities';

import { MyContext, UserInput, UserResponse } from '../types';

const resolvers: IResolvers = {
  /************************* Mutations *************************/

  Mutation: {
    // Register a user
    register: async (
      _: void,
      input: UserInput,
      { em }: MyContext
    ): Promise<UserResponse> => {
      if (input.username.trim().length < 3)
        return {
          errors: [
            {
              field: 'Username',
              message: 'Username must be at least 3 characters!',
            },
          ],
        };

      if (input.password.length < 8)
        return {
          errors: [
            {
              field: 'Password',
              message: 'Password must be at least 8 characters!',
            },
          ],
        };

      const hashedPass = await argon2.hash(input.password);
      const user = em.create(User, {
        username: input.username,
        password: hashedPass,
      });

      try {
        await em.persistAndFlush(user);
      } catch (err) {
        if (err.code === '23505') {
          return {
            errors: [
              {
                field: 'Username',
                message: 'Username already exists!',
              },
            ],
          };
        }
      }

      return {
        user,
      };
    },

    // Log in a user
    login: async (
      _: void,
      input: UserInput,
      { em }: MyContext
    ): Promise<UserResponse> => {
      const user = await em.findOne(User, {
        username: input.username.toLowerCase(),
      });

      if (!user)
        return {
          errors: [
            {
              field: 'User',
              message: 'Incorrect Login Information!',
            },
          ],
        };

      const validPass = await argon2.verify(user.password, input.password);

      if (!validPass)
        return {
          errors: [
            {
              field: 'User',
              message: 'Incorrect Login Information!',
            },
          ],
        };

      return {
        user,
      };
    },
  },

  /*************************  *************************/
};

export default resolvers;
