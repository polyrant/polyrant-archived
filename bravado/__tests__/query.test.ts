import { ApolloServer } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import { buildSchema } from 'type-graphql';

import { PingResolver } from '../src/resolvers/ping';

const PING = gql`
  {
    ping
  }
`;

const server = async () => {
  return new ApolloServer({
    schema: await buildSchema({
      resolvers: [PingResolver],
    }),
    mockEntireSchema: false,
    mocks: true,
  });
};

describe('queries', () => {
  test('ping', async () => {
    const { query } = createTestClient(await server());

    const res = await query({ query: PING });
    expect(res).toMatchSnapshot();
  });
});
