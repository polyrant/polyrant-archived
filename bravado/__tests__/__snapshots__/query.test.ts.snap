// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries ping 1`] = `
Object {
  "data": Object {
    "ping": "Pong!",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
